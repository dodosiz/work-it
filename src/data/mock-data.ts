import { AppState } from "./store";

export const DATA: AppState = {
	notifications: {
		message: undefined,
	},
	tasksState: {
		tasks: [
			{
				id: "task1",
				assignee: {
					id: "user1",
					firstName: "Alejandro",
					lastName: "Matthews",
					role: "Project manager",
				},
				title: "Plan the requirements for the next release.",
				dateFinished: undefined,
				description: "",
			},
			{
				id: "task2",
				assignee: {
					id: "user1",
					firstName: "Alejandro",
					lastName: "Matthews",
					role: "Project manager",
				},
				title: "Meet with the stakeholders to discuss the new requirements.",
				dateFinished: undefined,
				description: "",
			},
			{
				id: "task3",
				assignee: {
					id: "user2",
					firstName: "Mai",
					lastName: "Yoder",
					role: "UI/UX designer",
				},
				title: "Design the mocks for the new user interface.",
				dateFinished: undefined,
				description: "",
			},
			{
				id: "task4",
				assignee: {
					id: "user7",
					firstName: "Erik",
					lastName: "Ferguson",
					role: "Business analyst",
				},
				title: "Analyse the dependencies between the upcomming requirements.",
				dateFinished: undefined,
				description: "",
			},
		],
		taskFormOpened: false,
	},
	usersState: {
		userFormOpened: false,
		userIdToEdit: undefined,
		users: [
			{
				id: "user1",
				firstName: "Alejandro",
				lastName: "Matthews",
				role: "Project manager",
			},
			{
				id: "user2",
				firstName: "Mai",
				lastName: "Yoder",
				role: "UI/UX designer",
			},
			{
				id: "user3",
				firstName: "Wade",
				lastName: "Delgado",
				role: "Marketing",
			},
			{
				id: "user4",
				firstName: "Angeline",
				lastName: "Shea",
				role: "Senior front end developer",
			},
			{
				id: "user5",
				firstName: "Claudine",
				lastName: "Martin",
				role: "Section back end developer",
			},
			{
				id: "user6",
				firstName: "Cruz",
				lastName: "Russo",
				role: "DevOps",
			},
			{
				id: "user7",
				firstName: "Erik",
				lastName: "Ferguson",
				role: "Business analyst",
			},
			{
				id: "user8",
				firstName: "Charles",
				lastName: "Harrell",
				role: "Data analyst",
			},
			{
				id: "user9",
				firstName: "Lonnie",
				lastName: "Wade",
				role: "Security engineer",
			},
			{
				id: "user10",
				firstName: "Cassie",
				lastName: "Chung",
				role: "Quality assurance",
			},
			{
				id: "user11",
				firstName: "Dana",
				lastName: "Montes",
				role: "Junior developer",
			},
			{
				id: "user12",
				firstName: "Ladonna",
				lastName: "Ponce",
				role: "Junior developer",
			},
		],
	},
};
