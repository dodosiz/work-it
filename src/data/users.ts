import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { AppState } from "./store";
import { v4 } from "uuid";

export interface UsersState {
	users: User[];
}

export interface User {
	id: string;
	firstName: string;
	lastName: string;
	role: string;
}

interface AddUserPayload {
	firstName: string;
	lastName: string;
	role: string;
}

const initialState: UsersState = {
	users: []
};

const usersSlice = createSlice({
	name: "users",
	initialState,
	reducers: {
		addUser: (state, action: PayloadAction<AddUserPayload>) => {
			state.users.push({
				id: v4(),
				...action.payload
			});
		}
	}
});

export const { addUser } = usersSlice.actions;
export const usersReducer = usersSlice.reducer;

export const usersSelector = (state: AppState) => state.usersState.users;
