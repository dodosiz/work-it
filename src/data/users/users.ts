import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { AppState } from "../store";
import { v4 } from "uuid";

export interface UsersState {
	users: User[];
	userIdToEdit: string | undefined;
	userFormOpened: boolean;
}

export interface User {
	id: string;
	firstName: string;
	lastName: string;
	role: string;
}

interface AddUserPayload {
	firstName: string;
	lastName: string;
	role: string;
}

interface UpdateUserPayload {
	id: string;
	firstName: string;
	lastName: string;
	role: string;
}

interface UserIdPayload {
	id: string;
}

const initialState: UsersState = {
	users: [],
	userIdToEdit: undefined,
	userFormOpened: false,
};

const usersSlice = createSlice({
	name: "users",
	initialState,
	reducers: {
		addUser: (state, action: PayloadAction<AddUserPayload>) => {
			state.users.push({
				id: v4(),
				...action.payload,
			});
		},
		deleteUser: (state, action: PayloadAction<UserIdPayload>) => {
			const index = state.users.findIndex(
				(user) => user.id === action.payload.id
			);
			state.users.splice(index, 1);
		},
		editUser: (state, action: PayloadAction<UserIdPayload>) => {
			state.userIdToEdit = action.payload.id;
		},
		updateUser: (state, action: PayloadAction<UpdateUserPayload>) => {
			const index = state.users.findIndex(
				(user) => user.id === action.payload.id
			);
			state.users.splice(index, 1, action.payload);
		},
		openUserForm: (state) => {
			state.userFormOpened = true;
		},
		closeUserForm: (state) => {
			state.userFormOpened = false;
		},
	},
});

export const {
	addUser,
	deleteUser,
	editUser,
	updateUser,
	openUserForm,
	closeUserForm,
} = usersSlice.actions;
export const usersReducer = usersSlice.reducer;

export const usersSelector = (state: AppState) => state.usersState.users;
export const userFormOpenedSelector = (state: AppState) =>
	state.usersState.userFormOpened;
export const editedUserSelector = (state: AppState) => {
	return state.usersState.users.find(
		(u) => u.id === state.usersState.userIdToEdit
	);
};
