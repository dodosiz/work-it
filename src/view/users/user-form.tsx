import * as React from "react";
import { Button, Form } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import { addNotification } from "../../data/notifications/notifications";
import {
	addUser,
	closeUserForm,
	editedUserSelector,
	updateUser,
} from "../../data/users/users";

export function UserForm() {
	const dispatch = useDispatch();
	const editedUser = useSelector(editedUserSelector);
	const [firstName, setFirstName] = React.useState(
		editedUser ? editedUser.firstName : ""
	);
	const [lastName, setLastName] = React.useState(
		editedUser ? editedUser.lastName : ""
	);
	const [role, setRole] = React.useState(editedUser ? editedUser.role : "");
	const handleFirstNameChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setFirstName(event.target.value);
	};
	const handleLastNameChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setLastName(event.target.value);
	};
	const handleRoleChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setRole(event.target.value);
	};
	const handleSubmit: React.FormEventHandler<HTMLFormElement> = (event) => {
		event.preventDefault();
		if (editedUser) {
			dispatch(updateUser({ id: editedUser.id, firstName, lastName, role }));
			dispatch(
				addNotification({
					message: `Updated user ${firstName} ${lastName}.`,
				})
			);
		} else {
			dispatch(addUser({ firstName, lastName, role }));
			dispatch(
				addNotification({
					message: `Added user ${firstName} ${lastName}.`,
				})
			);
		}
		dispatch(closeUserForm());
	};
	const handleCancel = () => {
		dispatch(closeUserForm());
	};
	const submitDisabled = () => {
		return !(firstName.length && lastName.length && role.length);
	};
	return (
		<Form onSubmit={handleSubmit}>
			<Form.Group className="mb-3" controlId="firstName">
				<Form.Label>First name:</Form.Label>
				<Form.Control
					type="text"
					value={firstName}
					onChange={handleFirstNameChange}
				/>
			</Form.Group>
			<Form.Group className="mb-3" controlId="lastName">
				<Form.Label>Last name:</Form.Label>
				<Form.Control
					type="text"
					value={lastName}
					onChange={handleLastNameChange}
				/>
			</Form.Group>
			<Form.Group className="mb-3" controlId="role">
				<Form.Label>Role:</Form.Label>
				<Form.Control type="text" value={role} onChange={handleRoleChange} />
			</Form.Group>
			<Button disabled={submitDisabled()} variant="primary" type="submit">
				Save
			</Button>{" "}
			<Button variant="secondary" onClick={() => handleCancel()}>
				Cancel
			</Button>
		</Form>
	);
}
