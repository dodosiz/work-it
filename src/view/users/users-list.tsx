import * as React from "react";
import { Table, Button } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import {
	deleteUser,
	editUser,
	openUserForm,
	User,
	usersSelector,
} from "../../data/users/users";
import { BsFillTrashFill, BsPencilFill } from "react-icons/bs";
import { addNotification } from "../../data/notifications/notifications";

export function UsersList() {
	const dispatch = useDispatch();
	const users = useSelector(usersSelector);
	const handleDelete = (user: User) => {
		dispatch(deleteUser({ id: user.id }));
		dispatch(
			addNotification({
				message: `Deleted user ${user.firstName} ${user.lastName}.`,
			})
		);
	};
	const handleEdit = (id: string) => {
		dispatch(editUser({ id }));
		dispatch(openUserForm());
	};
	return (
		<Table striped bordered hover>
			<thead>
				<tr>
					<th>#</th>
					<th>First Name</th>
					<th>Last Name</th>
					<th>Role</th>
					<th>Action</th>
				</tr>
			</thead>
			<tbody>
				{users.map((user, index) => (
					<User
						key={user.id}
						user={user}
						index={index}
						handleEdit={handleEdit}
						handleDelete={handleDelete}
					/>
				))}
			</tbody>
		</Table>
	);
}

interface UserProps {
	index: number;
	user: User;
	handleEdit(id: string): void;
	handleDelete(user: User): void;
}

function User(props: UserProps) {
	return (
		<tr>
			<td>{props.index + 1}</td>
			<td>{props.user.firstName}</td>
			<td>{props.user.lastName}</td>
			<td>{props.user.role}</td>
			<td>
				<Button
					data-testid={`edit-button-${props.index}`}
					variant="outline-primary"
					onClick={() => props.handleEdit(props.user.id)}
				>
					<BsPencilFill />
				</Button>{" "}
				<Button
					data-testid={`delete-button-${props.index}`}
					variant="outline-danger"
					onClick={() => props.handleDelete(props.user)}
				>
					<BsFillTrashFill />
				</Button>
			</td>
		</tr>
	);
}
