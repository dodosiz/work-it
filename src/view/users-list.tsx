import * as React from "react";
import { useSelector } from "react-redux";
import { User, usersSelector } from "../data/users";

export function UsersList() {
	const users = useSelector(usersSelector);
	return <ul>
		{users.map((user, index) => (
			<User key={`$${user.name}-${index}`} user={user}/>
		))}
	</ul>;
}

interface UserProps {
	user: User;
}

function User(props: UserProps) {
	return <li>{props.user.name}, role: {props.user.role}</li>;
}
