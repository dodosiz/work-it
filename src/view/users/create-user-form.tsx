import * as React from "react";
import { Button, Form } from "react-bootstrap";
import { useDispatch } from "react-redux";
import { addUser, closeUserForm } from "../../data/users";

export function CreateUserForm() {
	const dispatch = useDispatch();
	const [firstName, setFirstName] = React.useState("");
	const [lastName, setLastName] = React.useState("");
	const [role, setRole] = React.useState("");
	const handleFirstNameChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setFirstName(event.target.value);
	};
	const handleLastNameChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setLastName(event.target.value);
	};
	const handleRoleChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setRole(event.target.value);
	};
	const handleSubmit: React.FormEventHandler<HTMLFormElement> = (event) => {
		event.preventDefault();
		dispatch(addUser({ firstName, lastName, role }));
		dispatch(closeUserForm());
	};
	const submitDisabled = () => {
		return !(firstName.length && lastName.length && role.length);
	};
	return (
		<Form onSubmit={handleSubmit}>
			<Form.Group className="mb-3" controlId="firstName">
				<Form.Label>First name:</Form.Label>
				<Form.Control
					type="text"
					value={firstName}
					onChange={handleFirstNameChange}
				/>
			</Form.Group>
			<Form.Group className="mb-3" controlId="lastName">
				<Form.Label>Last name:</Form.Label>
				<Form.Control
					type="text"
					value={lastName}
					onChange={handleLastNameChange}
				/>
			</Form.Group>
			<Form.Group className="mb-3" controlId="role">
				<Form.Label>Role:</Form.Label>
				<Form.Control type="text" value={role} onChange={handleRoleChange} />
			</Form.Group>
			<Button disabled={submitDisabled()} variant="primary" type="submit">
				Submit
			</Button>
		</Form>
	);
}
