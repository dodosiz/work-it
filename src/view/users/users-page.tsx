import * as React from "react";
import { useDispatch, useSelector } from "react-redux";
import {
	closeUserForm,
	editedUserSelector,
	userFormOpenedSelector,
} from "../../data/users/users";
import { ModalDialog } from "../modal/modal-dialog";
import { UserForm } from "./user-form";
import { UsersList } from "./users-list";

export function UsersPage() {
	const dispatch = useDispatch();
	const userFormOpened = useSelector(userFormOpenedSelector);
	const editedUser = useSelector(editedUserSelector);
	const handleCloseUserForm = () => {
		dispatch(closeUserForm());
	};
	return (
		<div>
			<ModalDialog
				handleClose={handleCloseUserForm}
				show={userFormOpened}
				title={editedUser ? "Edit user" : "Create new user"}
			>
				<UserForm />
			</ModalDialog>
			<UsersList />
		</div>
	);
}
