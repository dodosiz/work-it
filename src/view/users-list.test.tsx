import { configureStore } from "@reduxjs/toolkit";
import * as React from "react";
import { Provider } from "react-redux";
import { usersReducer } from "../data/users";
import { UsersList } from "./users-list";
import {render} from "@testing-library/react";

describe("Users list component:", () => {
	test("displays a list of users", () => {
		const store = configureStore({
			reducer: {
				usersState: usersReducer
			},
			preloadedState: {
				usersState: {
					users: [
						{name: "George", role: "developer"},
						{name: "Nick", role: "graphic designer"}
					]
				}
			}
		});
		const component = render(<Provider store={store}>
			<UsersList />
		</Provider>,
		);
		expect(component.container).toMatchSnapshot();
	});
});
