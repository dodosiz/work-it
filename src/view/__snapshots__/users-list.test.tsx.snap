// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users list component: displays a list of users 1`] = `
<div>
  <ul>
    <li>
      George
       
      Daniels
       
      <i>
        role: 
        developer
      </i>
    </li>
    <li>
      Nick
       
      Gates
       
      <i>
        role: 
        graphic designer
      </i>
    </li>
  </ul>
</div>
`;
